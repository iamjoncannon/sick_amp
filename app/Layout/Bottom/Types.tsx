export interface SongObject {
    ID: number 
    ALBUM: string 
    TITLE: string 
    ARTIST: string 
    TRACKNUMBER: string
}

export interface songStateType {
    [ key: number ] : SongObject
}

export interface Playlist {
    Title: string
    id: number
    ids: number[]
}

export interface Store {
    SelectedPlaylist: number
    PlayLists : Playlist[] | void 
    Songs: songStateType | void 
    ColumnHash : { [ Column : string] : Boolean} | void 
}

