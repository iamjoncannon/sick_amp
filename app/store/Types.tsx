export interface SongObject {
    ID: number 
    ALBUM: string 
    TITLE: string 
    ARTIST: string 
    TRACKNUMBER: string
    FILENAME: string
}

export interface songStateType {
    [ key: number ] : SongObject
}

export interface Playlist {
    Title: string
    id: number
    ids: number[]
}

export interface PlaylistState {
    [key : number] : Playlist
}

export interface Transport {
    previous: number | void 
    current: number | void
    next: number | void
}

export interface Store {
    Transport : Transport 
    isPlaying: Boolean
    SelectedPlaylist: number | string | void
    RunningPlaylist: number | string | void
    PlayLists : PlaylistState | void 
    Songs: songStateType | void 
    ColumnHash : { [ Column : string] : Boolean} | void 
    draggedOverPlaylist: Boolean    
}

