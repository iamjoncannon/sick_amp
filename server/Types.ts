export interface SongObject {
    ID: number 
    ALBUM: string 
    TITLE: string 
    ARTIST: string 
    TRACKNUMBER: string
    FILENAME? : string
}

export interface songStateType {
    [ key: number ] : SongObject
}

export interface Playlist {
    Title: string
    id: number
    ids: number[]
}

export interface returnDataType {
    Songs: songStateType
    PlayLists: Playlist[]
}
