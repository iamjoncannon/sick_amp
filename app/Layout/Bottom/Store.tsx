import React from 'react';
import * as Types from './Types'

const initialState : Types.Store = {
    isHydrated : false,
    SelectedPlaylist: 1,
    PlayLists: null,
    Songs: null,
};

interface ReduxAction {
    type: string
    payload: string
}

function reducer(state : Types.Store, action : ReduxAction ) {

    switch (action.type) {

        case 'HYDRATE': {

            return {...state, Songs: action.payload, isHydrated : true }
        }

        case 'SELECT_PLAYLIST':
        
            return { ...state, SelectedPlaylist: action.payload }
        
        case 'ADD_SONG_TO_PLAYLIST':

            return {...state}
        
        default:
          return state;
      }
}

export const Store = React.createContext(null);

export function StoreProvider(props : any) {

    const [state, dispatch] = React.useReducer(reducer, initialState);
    const value = { state, dispatch };

    return (
        <Store.Provider value={value}>
            {props.children}
        </Store.Provider>
    )
}
