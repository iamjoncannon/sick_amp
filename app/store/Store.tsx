import React from 'react';
import * as Types from './Types'

const initialState : Types.Store = {
    Transport : {
        current: 1, 
    },
    isPlaying: false,
    SelectedPlaylist: "All",
    RunningPlaylist: "All",
    Columns: null,
    PlayLists: null,
    Songs: null,
    draggedOverPlaylist: null,
    token: "dev"
};

interface ReduxAction {
    type: string
    // ideally this would be an enum 
    // with the contract for every action
    payload: any 
}

function reducer(state : Types.Store, action : ReduxAction ) {

    console.log("Action: ", action.type, action.payload)

    switch (action.type) {

        case 'HYDRATE_PLAYLISTS': {

            return {...state, PlayLists: action.payload}
        }

        case 'HYDRATE_SONGS': {

            const { PlayList, Page, data } = action.payload

            // pouring songs into the pool

            const Songs = {...state.Songs, ...data}

            // update the hydration hash for the playlist 

            const PlayLists = {...state.PlayLists}

            PlayLists[PlayList].hydrated[Page] = true 
  
            return {...state, PlayLists : PlayLists, Songs: Songs }
        }

        case 'PLAYLIST_END': {

            const { PlayList, Page } = action.payload

            const next_PlayLists = {...state.PlayLists}

            next_PlayLists[PlayList].hydrated["Complete"] = true 
            next_PlayLists[PlayList].hydrated[Page] = true 

            return {...state, PlayLists: next_PlayLists }
        }

        case 'HYDRATE_COLUMNS': {

            return {...state, Columns: action.payload.data}
        }
        

        case 'SELECT_PLAYLIST':
        
            return { ...state, SelectedPlaylist: action.payload }

        case 'PLAY_TRACK': {

            const current : number = Number(action.payload)

            const { SelectedPlaylist } = state 

            let newTransport : Types.Transport = { current } 

            return {...state, Transport: newTransport, isPlaying: true, RunningPlaylist: SelectedPlaylist}
        }
        
        case "PLAY_PREVIOUS_TRACK": {

            const { RunningPlaylist, PlayLists, Transport : { current } } = state 

            // architecture note- I decided to structure the "All" Playlist and the specific
            // playlists differently- see ./Types -- this requires us to handle each differently
            // in the playlist actions - this could certainly be revisited and refactored 

            // if its playing from the all playlist, then its simply the previous track
            // or the end of the playlist if current = 0 

            let next_current 

            if(RunningPlaylist === "All"){

                next_current = current === 0 ? PlayLists[RunningPlaylist].length -1 : Number(current) - 1 
            }
            else{

                // if its a specific playlist, then we need to find the index of the track in the 
                // playlists ids and return the previous index, or end if 0 

                let CurrentPlaylist = PlayLists[RunningPlaylist].ids

                const current_index = CurrentPlaylist.indexOf(current)

                next_current = current_index === 0 ? CurrentPlaylist[CurrentPlaylist.length -1] : CurrentPlaylist[current_index -1 ] 
            
            }

            let newTransport : Types.Transport = { current: next_current } 
            
            return {...state, Transport: newTransport, isPlaying: true}
        }

        case "PLAY_NEXT_TRACK": {

            const { RunningPlaylist, PlayLists, Transport : { current } } = state 

            let next_current 
            let CurrentPlaylist 

            if(RunningPlaylist === "All"){

                CurrentPlaylist = PlayLists[RunningPlaylist]

                next_current = current === CurrentPlaylist.length -1 ? 0 : Number(current) + 1 
            }
            else{

                CurrentPlaylist = PlayLists[RunningPlaylist].ids

                const current_index = CurrentPlaylist.indexOf(current)

                next_current = current_index === CurrentPlaylist.length -1 ? CurrentPlaylist[0] : CurrentPlaylist[current_index + 1 ] 
            }

            let newTransport : Types.Transport = { current: next_current } 
            
            return {...state, Transport: newTransport, isPlaying: true}
        }

        case 'ADD_SONG_TO_PLAYLIST':{

            const { PlayLists } = state 
            const { payload: { song, playlist } } = action

            const nextPlaylists = [...PlayLists]

            if(!nextPlaylists[playlist].ids.includes(Number(song))){

                nextPlaylists[playlist].ids = [...PlayLists[playlist].ids, Number(song)]
            }

            return {...state, PlayLists : nextPlaylists, draggedOverPlaylist: null}
        }

        case "TOGGLE_PLAYERSTATE": {

            return {...state, isPlaying: !state.isPlaying}
        }


        case "REARRANGE_PLAYLIST": {

            const { item_to_put_before, item_to_be_moved } = action.payload

            let next_playlist_object = [...state.PlayLists]
            
            if(state.SelectedPlaylist !== "All"){

                let targetPlaylist = next_playlist_object[state.SelectedPlaylist].ids

                let nextTargetPlaylist = []

                // this is not the most clever way to do this
                // but it works 
                for(let i = 0; i < targetPlaylist.length; i++){

                    if(targetPlaylist[i] !== item_to_be_moved){
                        
                        if(targetPlaylist[i] === item_to_put_before){

                            nextTargetPlaylist.push(item_to_be_moved)
                        }

                        nextTargetPlaylist.push(targetPlaylist[i])
                    }
                }

                next_playlist_object[state.SelectedPlaylist].ids = nextTargetPlaylist
                
            }
            
            // this could be refactored- playlists is an object
            // we also want to treat like an array - "All" needs
            // to be reappended if we destructure as above 
                
            next_playlist_object["All"] = [...Object.values(state.Songs)]
                        
            return {...state, PlayLists : next_playlist_object}
        }

        case "ADD_PLAYLIST":{

            let next_playlist_object = {...state.PlayLists}

            next_playlist_object[next_playlist_object.length] = { name: "New PlayList", id: next_playlist_object.length, files: [ ]}

            return {...state, PlayLists : next_playlist_object}
        }

        case "DRAG_OVER_PLAYLIST": {

            return {...state, draggedOverPlaylist: action.payload}
        }
        
        default:
          return state;
      }
}

export const Store = React.createContext(null);

export function StoreProvider(props : any) {

    const [state, dispatch] = React.useReducer(reducer, initialState);
    const value = { state, dispatch };

    return (
        <Store.Provider value={value}>
            {props.children}
        </Store.Provider>
    )
}

function clone(item){

    return JSON.parse(JSON.stringify(item))
}