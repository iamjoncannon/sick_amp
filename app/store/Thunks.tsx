const { default: axios } = require('axios');
import { api_url } from '../../config'

// called by PlayLists 

export async function fetchInitialData(token: string, dispatch: any){

    axios.get(`${api_url}/folders/?api_key=${token}`).then(( { data } )=>{
    
        const formattedData = {}
        
        for(let playlist in data){

            formattedData["All"] = { name: "All Songs", hydrated: {}  }
            
            // make sure the local hash key is the same 
            // as the id returned from the server
            formattedData[data[playlist].id] = data[playlist]
            
            // add hydrated boolean to prevent manage
            // fetching of data for each playlist 
            formattedData[data[playlist].id].hydrated = {}
        }

        dispatch({
            type: "HYDRATE_PLAYLISTS",
            payload: formattedData
        })
        
    }).catch(err=>console.log(err))

    hydrateSongs(token, dispatch, "All", 1)
}

// called by EditablePlayList

export async function postPlayList(name: string, token: string, dispatch: any){

    let result

    try{

        result = await axios.post(`${api_url}/folders/?api_key=${token}`, { name })
    }
    catch(err){

        console.log(err)
        return 
    }

    dispatch({type: 'UPDATE_PLAYLISTS', payload: result.data})
}

export async function putPlayList(name: string, token: string, dispatch: any){

    const endpoint = `${api_url}/folders/?api_key=${token}`
    let result
    
    try{
        
        result = await axios.put(endpoint, { name })
    }
    catch(err){

        console.log(err)
        dispatch({type: "CANCEL_UPDATE_PLAYLISTS"})
        return 
    }
    
    dispatch({type: 'UPDATE_PLAYLISTS', payload: result.data})
}

// addSongToPlaylist POST /files/{file_id}/folders/{folder_id}

export async function addSongToPlaylist(folder_id: number, song_id: number, token: string, dispatch: any){

    const endpoint = `${api_url}/files/${song_id}/folders/${folder_id}?api_key=${token}`

    let result 

    try {

        result = await axios.post(endpoint)
    }
    catch(err){

        console.log("addSongToPlaylist thunk error: ", err)
        return
    }

    console.log(result)


    /*

    dispatch({
                type: "ADD_SONG_TO_PLAYLIST",
                payload: {
                    newPlayListObject = result.data,
                    playlist_id: folder_id
                }
            })

    */

}


export async function rearrange_playlist(folder_id: number, song_id: number, position: number, token: string, dispatch: any){

    // files/{fileid}/folders/{folderid} { position: int }
    const endpoint = `${api_url}/files/${song_id}/folders/${folder_id}?api_key=${token}`
    
    let result
    
    try{
        
        result = await axios.put(endpoint, { position })
    }
    catch(err){

        console.log(err)
        dispatch({type: "CANCEL_UPDATE_PLAYLISTS"})
        return 
    }

    dispatch({
        type: "REARRANGE_PLAYLIST",
        payload: {
            
            updatedPlaylist: result.data
        }
    })

}


// called by Songs 

export async function hydrateColumns(token: string, dispatch: any){

    let data 

    try{
        data = await axios.get(`${api_url}/fields?api_key=${token}`)
    }
    catch(err){
        console.log(err)
    }

    dispatch({
        type: "HYDRATE_COLUMNS",
        payload: data
    })

}

// called by SongTable 

export async function hydrateSongs(token: string, dispatch: any, PlayList: string, Page: number){

    axios.get(`${api_url}/files/?page=${Page}&api_key=${token}`).then(( { data } )=>{
    
        const formattedData = {}

        for(let song in data){

            const { id } = data[song]

            formattedData[id] = data[song]
        }

        if(Object.keys(formattedData).length === 0){

            dispatch({
                type: "PLAYLIST_END",
                payload: { PlayList, Page }
            })
        }
        else{
            
            dispatch({
                type: "HYDRATE_SONGS",
                payload: { PlayList: PlayList, Page, data: formattedData }
            })
        }

    }).catch(err=>console.log(err))
}



