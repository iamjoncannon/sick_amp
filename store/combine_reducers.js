import { combineReducers } from 'redux'
import User_state from './User/reducer_for_User'

export default combineReducers({
 User_state,
})