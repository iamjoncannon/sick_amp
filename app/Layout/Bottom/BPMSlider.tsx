import React from 'react';
import { Store } from '../../store/Store'
import styled from 'styled-components'
import Slider from 'rc-slider';
import 'rc-slider/assets/index.css';

export const SliderContainer = styled.div`

    width: 80%;
    margin: 1vh;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;

    .range {
        display: flex;
        flex-direction: row;
        justify-content: space-between;
        width: 100%;
        font-size: 1.75vh;
    }
    
    div .rc-slider-rail {

        background-color: ${props => props.theme.fontColor};
    }

    div .rc-slider-track {

        background-color: ${props => props.theme.logoColor};
    }

    div .rc-slider-handle {

        background-color: ${props => props.theme.logoColor};
        border: ${props => props.theme.logoColor};

    }

    .rc-slider-handle:active{
        box-shadow: 0 0 1px ${props => props.theme.logoColor};
    }

    .rc-slider-handle:focus{
        box-shadow: 0 0 1px ${props => props.theme.logoColor};
    }
`

const BPMSlider = () => {

    const whichStore = process.env.NODE_ENV === 'test' ? require('./BPMSlider.test').MockStore : Store ;
    
    const { state, dispatch } = React.useContext(whichStore);
    const { FilterState: { bpm } } = state 
    const [ value, setValue ] = React.useState(bpm)

    const handleDispatch = () => {

        dispatch({type: "MUTATE_FILTERSTATE", payload: { field: "bpm", value }})
    }

    return (
        <SliderContainer>

            <span>Beats Per Minute</span>

            <Slider.Range 
                value={value} 
                allowCross={false} 
                min={50}
                max={200}
                defaultValue={[50, 200]} 
                onChange={(value)=>setValue(value)} 
                onAfterChange={()=>handleDispatch()}
            />

            <div className="range">
                <span>{value[0]}</span>
                <span>{value[1]}</span>
            </div>
        
        </SliderContainer>
    )
}

export default BPMSlider