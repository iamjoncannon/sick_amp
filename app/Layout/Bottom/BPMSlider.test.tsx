import React from 'react'
import { mount } from 'enzyme'
import BPMSlider, { SliderContainer } from './BPMSlider'
import Types from '../../store/Types'
import { MUTATE_FILTERSTATE } from '../../store/Store'
import { act } from 'react-dom/test-utils';

/*
    Intended behavior: 

    [x] displays BPM from FilterState in two spans 
    [x] onChange callback sets value property on local state
    [x] onAfterChange callback dispatches MUTATE_FILTERSTATE to store with local state

*/

export let MockStore 
let mockState 
let dispatchedActions = []
let mockInitialState

function mockReducer(state : Types.Store, action : ReduxAction ) {
    
    switch (action.type) {

        case "MUTATE_FILTERSTATE": {

            return MUTATE_FILTERSTATE(state, action)
        }
    }
}

describe("BPMSlider", ()=>{
    let nextState = [] 
    const setState = jest.fn();
    const useStateSpy = jest.spyOn(React, 'useState')
    useStateSpy.mockImplementation((init) => [init, setState]);

    beforeEach(() => {

        MockStore = React.createContext(null);

        mockInitialState = {

              FilterState: {
                  bpm: [100, 200]
              }          
        }
    });

    function MockStoreProvider(props : any) {

        let [ state, dispatch ] = React.useReducer(mockReducer, mockInitialState);

        mockState = state 

        let mockDispatch = function(...args){

            dispatchedActions.push(...args)

            dispatch(...args)

            nextState.push(state)
        }

        return (
            <MockStore.Provider value={{ state, dispatch: mockDispatch }}>
                {props.children}
            </MockStore.Provider>
        )
    }

    function TestComponent (prop){

        return mount(
            <MockStoreProvider>

                <BPMSlider />

            </MockStoreProvider>
            );
    }

    it('Component renders correctly', () => {

        const Tree = TestComponent()
        
        expect(Tree).toMatchSnapshot();
    });

    it("displays BPM from FilterState in two spans ", ()=>{

        const Tree = TestComponent()
        const spans = Tree.find('span')
        expect(spans.get(1).props.children).toEqual(mockInitialState.FilterState.bpm[0])
        expect(spans.get(2).props.children).toEqual(mockInitialState.FilterState.bpm[1])
    })
    
    it("onChange callback sets value property on local state", ()=>{

        const Tree = TestComponent()

        const Slider = Tree.find(SliderContainer).props().children[1]

        const newValue = [150,200]

        Slider.props.onChange(newValue)
        
        expect(setState.mock.calls[0][0]).toEqual(newValue)
    })

    it("onAfterChange callback dispatches MUTATE_FILTERSTATE to store with current value of input", ()=>{

        dispatchedActions = []

        const Tree = TestComponent()

        let Slider = Tree.find(SliderContainer).props().children[1]

        const { value } = Slider.props

        act(()=>{
            
            Slider.props.onAfterChange()
        })

        expect(dispatchedActions[0].type).toEqual("MUTATE_FILTERSTATE")
        expect(dispatchedActions[0].payload.value).toEqual(value)

        expect(nextState[0].FilterState.bpm).toEqual([100,200])
    })
})
