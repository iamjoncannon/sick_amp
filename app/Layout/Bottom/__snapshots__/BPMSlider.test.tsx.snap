// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BPMSlider Component renders correctly 1`] = `ReactWrapper {}`;
